### PR TITLE
Create Initial Parity Stability Module

### PR DESCRIPTION
**Pull request by dtribble**
refs: #4663

## Description

Initial cut at a Parity Stability Module: 
* provide an anchoring token (e.g., like USDC) to mint a local stable token (e.g., RUN), and vice versa
* charge different fees for trading into and out of the stable token
* enforce an external minting limit, by limiting the size of the pool of anchor

### Security Considerations

It must enforce a minting limit for stable.

### Testing Considerations

Key test cases:
* give `anchor`, get `stable`
  * fees are charged
  * the pool balance is updated
* give `stable`, get `anchor`
  * fees are charged
  * the pool balance is updated
* from a non-zero pool, give `anchor` to overflow the pool
  * offer is rejected
  * the pool is not changed

### CODE REVIEW

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js:119]: Why is an assert better than letting the reallocate fail? I think the assert is extraneous.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js:135]: reallocate will verify this.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]:
```suggestion
    const asStable = AmountMath.coerce(stableBrand, given);
```

If `given` isn't trustworthy (it is, Zoe has checked it) then `anchorAfterTrade`  should use asStable rather than given.  If `given` has the wrong brand, the arithmetic and reallocation will fail.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]: `vaultFactory/collectRewardFees` shows how it is made accessible to the fee collector.  Any reason not to incorporate that code now?

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]: This is a fine temporary stand-in for governance, but I think it ought to use governance before merging.

would you like someone to add governance?

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js]: If this used NatSafeMath and Ratios, I wouldn't worry that future tests might pass in `given`s that were small enough that this wouldn't be reliable.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js]: does Promise.all add anything here?
```suggestion
  const psmBundle = await makeBundle(psmRoot);
```

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]: I don't think I'd call this 'Liquidity'. I think 'collateral', 'balance', or 'reserves' would make more sense to me.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js]: I think once the test is working, this should be dropped.  (Unless we're going to give `trace` a toggle.)

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js:30]: Please add a top comment explaining the point of the contract. We usually also document the keywords in use and the main invitations. It will be much easier fro someone to add contract docs to our documentation site if we have at least a rudimentary description here.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js:53]: I think we also need a multiplication factor here. Even if we're trading at "par", the different currencies may have a different number of decimal places.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js]: use it or delete it.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js:217]: please uncomment.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js:180]: This tests trying to extract more than has been added. I also want to see trying to get more than the `mintLimit`

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:135]: I wanted to check before minting to avoid trivially causing mint/burn efforts.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: I like 'balance' and 'reserve'. It might make sense to change all the "anchor" names to "reserve" as well, since that's the reserve asset for this contract.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js]: It's to keep the pattern from other tests that use multiple bundles. But can be deleted.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js:180]: It's contributing more than `mintLimit` in `anchor` (line 168), so that is what it's testing.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js:217]: This is related to the TODO.  I would love to uncomment, but then it fails because `getAmountOf` fails internally. That's because `t.Out` is undefined (as in line 178). I'm leaving the comment until you and Mark see that and agree that it's expected. Because it surprised me!

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:119]: This is just because it was paralleling the other branch, which used the assert to avoid a spurious minting event.

Comment (turadg) [reply packages/run-protocol/src/psm/psm.js:119]: Either way, WDYT of a `@throws` comment on the function jsdoc to describe?

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: This is doing a 1:1 conversion from USDC to RUN. Your point earlier that this needs to use a price even if it is 1:1 because they might have different decimals is spot on though.  So this needs to change to a multiply.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: Will do.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: Governance APIs are still moving. I don't think it should be blocked on that. It will need that before it could be turned on, but there's multiple concurrent streams of work that could continue on this (governance and bootstrap and making it durable).

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:119]: I shall avoid @throws both because of Java PTSD (pervasive throw system disorders) and because we are still debating exactly how to abruptly terminate.

File: packages/run-protocol/src/collectFees.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/collectFees.js]:
```suggestion
 * @returns {{makeCollectFeesInvitation: () => Invitation}}
```

File: packages/run-protocol/src/collectFees.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/collectFees.js:19]: Didn't quite get rid of all the `RUN`. Should this be `[keyword]` as well?

File: packages/run-protocol/src/makeTracer.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/makeTracer.js:19]: If you don't expect info and default to diverge, these should use the same code. If you do expect them to diverge, how would they differ?

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]: I think this comment is now extraneous.

File: packages/run-protocol/test/psm/test-psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/psm/test-psm.js:258]: I don't think it makes sense to leave `trace` statements in tests. If they're important enough to leave around, they should be converted to assertions.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:53]: I added it to terms rather than governance

File: packages/run-protocol/src/collectFees.js
Comment (dtribble) [reply packages/run-protocol/src/collectFees.js:19]: No I was deliberately taking from an application-specified keyword and dumping them into the 'RUN' keyword on the rewards seat so that all rewards end up on the same keyword (regardless of the internal keyword used).  Just solving one problem at a time.

File: packages/run-protocol/src/makeTracer.js
Comment (dtribble) [reply packages/run-protocol/src/makeTracer.js:19]: `info` uses `console.info`; `default` uses `console.log`. So Info I believe is normally not on.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js:258]: They should be using the logging that only turns on if the test fails. They aren't asserts. They are informational.

File: packages/run-protocol/src/makeTracer.js
Comment (turadg) [reply packages/run-protocol/src/makeTracer.js:19]: > So Info I believe is normally not on

It's often disabled but Chrome, for example, defaults `info` to on:
Screen Shot 2022-03-24 at 4 15 49 PM

> info uses console.info; default uses console.log

In Node, console.info is an alias for console.log.  ([ref](https://nodejs.org/api/console.html#consoleinfodata-args))

Maybe this will all get sorted out in the emerging [Console Standard](https://console.spec.whatwg.org/#loglevel-severity).

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: I believe that it should not finish governance before merging. Governance raises additional questions, tests cases, bootstrap integration issues, etc. Similarly, it should not wait for integrating durability. Those are orthogonal features that can be integrated independently.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/psm/psm.js]: The comment is no longer needed true, since we're using collectFeesInvitation.

File: packages/run-protocol/src/collectFees.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/collectFees.js]: updating this type will be useful to callers.

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/psm/psm.js]: Were you going to change the names, or were you waiting for more discussioni?

File: packages/run-protocol/src/psm/psm.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/psm/psm.js]: Is this intended to prevent minting that adds up to more than mint limit, or just individual transactions that are larger than the limit? It seems to be doing the latter, and I'm not sure why that's useful. As a client, I can make multiple requests to stay under the limit.

File: packages/run-protocol/src/collectFees.js
Comment (turadg) [reply packages/run-protocol/src/collectFees.js]: We could also just drop the `@returns`. It's implied by what's actually returned.

File: packages/run-protocol/src/makeTracer.js
Comment (turadg) [inline packages/run-protocol/src/makeTracer.js:3]: if we're going to have this option, `enabled = false` makes more sense to me. Bigger change but this change is already out of scope ;-)

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js]: this is defined in 13 places. now may be a good time to pick a place to keep it.

also `10_000n` for readibility.

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js]: s/singe/single

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:38]: `AmountKeywordRecord` can have any number of keywords. Consider a tuple: `[Keyword, Amount]`

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:41]: If you are going to stick with an `AmountKeywordRecord` parameter, check that it has exactly one keyword/amount.

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:58]: Isn't `anchorBrand` term redundant with `anchorPerStable`? 
```suggestion
  const { anchorPerStable } = zcf.getTerms();
  const anchorBrand = anchorPerStable.numerator.brand;
```

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:64]: worth checking,
```
assert(keyEq(stableBrand, anchorPerStable.denominator.brand));
```

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [reply packages/run-protocol/src/psm/psm.js]: Fair. Though "HACK for simple governance API" sounds like it's got governance. Suggestion "XXX mock governance API for terms that will be governed"

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js]: too icy 🥶

s/specificy/specify

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [reply packages/run-protocol/src/psm/psm.js:119]: What's the status of resolving that debate?

When it's resolved, how will we know to come back and make this operate that way?

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js]: stray

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js]: stray. if this is meant to be left for future work, leave a comment. (though I think it's better to remove)

File: packages/run-protocol/src/collectFees.js
Comment (turadg) [inline packages/run-protocol/src/collectFees.js:10]: `makeMake` 🙄

**Review by turadg — COMMENTED** [review]
Took a pass at the code level, not economics.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js]: I switched to natSafeMath, but it seems on the edge of the obscurity being worse that not having the checking: 
```
    NatMath.floorDivide(
      NatMath.multiply(stable.value, NatMath.subtract(BASIS_POINTS, feeBP)),
      BASIS_POINTS,
    ),
```
vs
```
 BigInt((stable.value * (BASIS_POINTS - feeBP)) / BASIS_POINTS),
```

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: The git UI had several items visually collapsed so I missed them.  Fixed.

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js]: disabled them

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js:217]: @erights ^^

File: packages/run-protocol/test/psm/test-psm.js
Comment (dtribble) [reply packages/run-protocol/test/psm/test-psm.js:258]: `makeTracer` now supports disabling, so I did that.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: Now it's `getPoolBalance`

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: It prevents minting a total that gets to be more than the mint limit. It adds the incoming anchor amount to the total that's currently on the anchor pool, and asserts if that is over the limit. At least it's supposed to, which matches the first criteria.  I moved the addition into `assertUnderLimit` so it should be more clear now.

File: packages/run-protocol/src/makeTracer.js
Comment (dtribble) [reply packages/run-protocol/src/makeTracer.js:3]: That should be a separate work item, and only after letting people know at scrum.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:38]: It's for internal usage as a sanity check, so I wasn't trying to be too rigorous. It's more useful to have the same format as the arguments to decrementBy et al.  Since those APIs will be changing, I thought it not worth further effort yet.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:58]: The `anchorBrand` should be in `anchorPerStable`, but one is a ratio and the other a brand. From a terms point of view, I think it is more clear to have the brand be front and center.

Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:58]: Added an assert instead.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:64]:
```javascript
  assert(
    anchorPerStable.numerator.brand === anchorBrand &&
      anchorPerStable.denominator.brand === stableBrand,
    X`Ratio ${anchorPerStable} is not consistent with brands ${anchorBrand} and ${stableBrand}`,
  );
```

File: packages/run-protocol/src/collectFees.js
Comment (dtribble) [reply packages/run-protocol/src/collectFees.js:10]: Not my name. I just moved it. :)

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:119]: There's a separate discussion about 1. using `assert` vs. 2. use `assert` here instead of relying on reallocate. 
WRT 1: Vaults use asserts, so any change to that pattern will go through all of RUN protocol and change it.
WRT 2: I think that the details that includes fees will be informative in the easy mistake case that someone says "give 2 USDC want 2 RUN"

So I concluded that I should leave it.

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js:135]: An as above, I think that the details that includes fees will be informative in the easy mistake case that someone says "give 2 USDC want 2 RUN"

File: packages/run-protocol/src/psm/psm.js
Comment (dtribble) [reply packages/run-protocol/src/psm/psm.js]: It's not a ratio multiply. Much cleaner!

File: packages/run-protocol/src/makeTracer.js
Comment (turadg) [reply packages/run-protocol/src/makeTracer.js:3]: Fair point about backwards compatibility. Re: scrum, this is a `run-protocol` function so I think we have quorum on this review to make a decision.

I assume you're going to squash this PR into a single commit: `feat(run-protocol): initial Parity Stability Module`. Having this change in that commit is subpar hygiene but not a blocker.

File: packages/run-protocol/src/collectFees.js
Comment (turadg) [reply packages/run-protocol/src/collectFees.js:10]: sure, but if there's a better name this is a good time to change it.

It's common to have chained makers. is prepending "make" the pattern to follow? (I thought there must be something better but I can't think of one. I've asked in chat.)

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:21]: happy to see a @file comment. Seems like it would be even better at the top (before imports)

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [inline packages/run-protocol/src/psm/psm.js:13]: stray

File: packages/run-protocol/src/psm/psm.js
Comment (turadg) [reply packages/run-protocol/src/psm/psm.js:41]: Not sure if you saw this. I'll accept not changing the args because the APIs may be changing, but I think the code in master should have this check for what the API is now. Also conveys the expectations to the code reader.

```suggestion
  const fromValues = Object.values(txFrom);
  const toValues = Object.values(txTo);
  assert(fromValues.length === toValues.length === 1);
  assert(AmountMath.isEqual(fromValues[0], toValues[0]));
```

**Review by turadg — APPROVED** [review]
Some requests but none blocking.

**Review by turadg — COMMENTED** [review]
Just noticed this has auto-merge enabled so I'm changing to Comment review so you can see my requests before it's merged.

**Comment by turadg** [comment]
> Just noticed this has auto-merge enabled so I'm changing to Comment review so you can see my requests before it's merged.

TIL that adding a Comment review does not dismiss an Approve. So the auto-merge merged.

@dtribble please consider the comments added this morning.

**Pull request merged**
